File: ProblemSolving/450DSA/Java/src/main/java/org/redquark/techinterview/dsa/lists/ReverseALinkedListInGroupOfK.java
package org.redquark.techinterview.dsa.lists;

/**
 * @author Anirudh Sharma
 * <p>
 * Given a linked list of size N. The task is to reverse every k nodes (where k is an input to the function)
 * in the linked list.
 */
public class ReverseALinkedListInGroupOfK {

    private static Node reverseInGroupOfK(Node head, int k) {
        // Base condition
        if (head == null || k == 1) {
            return head;
        }
        // Dummy node
        Node dummy = new Node(-1);
        // Link dummy node to the head
        dummy.next = head;
        // Make dummy as the previous node
        Node previous = dummy;
        // Reference of the current node
        Node current = head;
        // Variable to store the number of nodes traversed
        int count = 0;
        // Loop through the list
        while (current != null) {
            count++;
            // Check if the count is divisible by k.
            // If it is then we need to reverse the sublist
            // between previous and current.next
            if (count % k == 0) {
                previous = reverse(previous, current.next);
                // Update the current node
                current = previous.next;
            }
            // If the count is not divisible by k then move
            // the pointer ahead
            else {
                current = current.next;
            }
        }
        return dummy.next;
    }

    private static Node reverse(Node start, Node end) {
        // Previous node of the current node
        Node previous = start.next;
        // Current node
        Node current = previous.next;
        // Next node of the current node
        Node next;
        // Loop for the whole interval
        while (current != end) {
            // Next node of the current node
            next = current.next;
            // Next of current will point to the previous
            current.next = start.next;
            // Current node will become the previous node
            start.next = current;
            // Move pointer ahead
            current = next;
        }
        previous.next = end;
        // Return head node of the reversed linked list
        return previous;
    }

    private static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " ");
            node = node.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        head.next.next.next.next = new Node(5);
        printList(reverseInGroupOfK(head, 2));

        printList(reverseInGroupOfK(null, 2));

        head = new Node(1);
        printList(reverseInGroupOfK(head, 3));
    }

    static class Node {
        final int data;
        Node next;

        Node(int data) {
            this.data = data;
        }
    }
}

File: ProblemSolving/450DSA/Python/src/lists/ReverseALinkedListInGroupOfK.py
"""
@author Anirudh Sharma

Given a linked list of size N. The task is to reverse every k nodes (where k is an input to the function)
in the linked list.
"""


class ListNode:
    def __init__(self, val=0, nextNode=None):
        self.val = val
        self.next = nextNode


def reverseKGroup(headNode: ListNode, k: int):
    # Base condition
    if headNode is None or k == 1:
        return headNode
    # Dummy node before headNode
    dummy = ListNode(-1)
    # Point the next of this dummy node to the current headNode
    dummy.next = headNode
    # Node to keep track of the previous node
    previous = dummy
    # Variable to keep count of the nodes in the linked list
    count = 0
    # Reference to the headNode which will be used to traverse
    current = headNode
    # Loop for all the nodes in the list
    while current is not None:
        count += 1
        if count % k == 0:
            previous = reverseList(previous, current.next)
            current = previous.next
        else:
            current = current.next
    return dummy.next


def reverseList(start, end):
    previous = start.next
    current = previous.next
    while current is not end:
        nextNode = current.next
        current.next = start.next
        start.next = current
        current = nextNode
    previous.next = end
    return previous


def printList(headNode):
    temp = headNode
    while temp is not None:
        print(temp.val, end=" ")
        temp = temp.next
    print()


if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    printList(reverseKGroup(head, 2))

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    printList(reverseKGroup(head, 3))

    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    printList(reverseKGroup(head, 1))

    head = ListNode(1)
    printList(reverseKGroup(head, 1))
